server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -665,10 +671,11 @@ private synchronized void consumeInternal(QuerySearchResult querySearchResult) {
             if (querySearchResult.isNull() == false) {
                 if (index == bufferSize) {
                     if (hasAggs) {
-                        ReduceContext reduceContext = aggReduceContextBuilder.forPartialReduction();
-                        InternalAggregations reducedAggs = InternalAggregations.topLevelReduce(Arrays.asList(aggsBuffer), reduceContext);
+                        InternalAggregations reducedAggs = InternalAggregations.topLevelReduce(
+                                Arrays.stream(aggsBuffer).map(Supplier::get).collect(toList()),
+                                aggReduceContextBuilder.forPartialReduction());
                         Arrays.fill(aggsBuffer, null);
-                        aggsBuffer[0] = reducedAggs;
+                        aggsBuffer[0] = () -> reducedAggs;
```

Comment:
should we nullify the rest of the array to make the reduced aggs eligible for gc ?

Reply:
++

Reply:
Actually the line right above does that.

Reply:
Right but we keep the serialized + deserialized form until after the partial reduce. We can try to release the serialized form early with:
````
 List<InternalAggregations> toReduce = Arrays.stream(aggsBuffer).map(Supplier::get).collect(toList());
 Arrays.fill(aggsBuffer, null);
 InternalAggregaions reducedAggs = InternalAggregations.topLevelReduce(toReduce, aggReduceContextBuilder.forPartialReduction());
 aggsBuffer[0] = () -> reducedAggs;
````
Or we can nullify the serialized form when the supplier is called like discussed below.

Reply:
Right! I noticed that right after I sent this. I'm playing with nulling the cell in the array as soon as I call `get`. That *feels* a little safer than nulling the bytes.